### PR TITLE
distribution: add on_top to validators' weights

### PR DESCRIPTION
**Pull request by foxpy**

### CODE REVIEW

File: contracts/strategy/src/contract.rs [hunk 1]
```diff
@@ -161,6 +161,7 @@ fn prepare_delegation_data(
         let delegation = drop_staking_base::msg::distribution::Delegation {
             valoper_address: validator.valoper_address.clone(),
             stake: validator_denom_delegation,
+            on_top: Uint128::zero(),
```
Comment (ratik) [inline contracts/strategy/src/contract.rs:164]: do I get it right that it's just a placeholder now?

Comment (foxpy) [reply contracts/strategy/src/contract.rs:164]: yes, it's just a placeholder for future, this PR only targets distribution contract

File: packages/base/src/msg/distribution.rs [hunk 2]
```diff
@@ -1,26 +1,231 @@
...
     pub total_weight: u64,
     pub delegations: Vec<Delegation>,
 }
+
+#[derive(Default)]
+pub struct StakeChanges {
+    changes: HashMap<String, Uint128>,
+}
+
+impl StakeChanges {
```
[the remark this answers is not in the document]
Comment (foxpy) [reply packages/base/src/msg/distribution.rs]: Unfortunately, I didn't figure out a clear way to define methods there and implement them in a contract, so I just abandoned the idea of using methods altogether. On the bright side, now thanks to test coverage, I found two uncovered code paths and tested them, too.

File: contracts/distribution/src/contract.rs [hunk 3]
```diff
@@ -38,517 +42,246 @@ pub fn query(_deps: Deps, _env: Env, msg: QueryMsg) -> ContractResult<Binary> {
...
-
-        assert_eq!(
-            distribution,
-            vec![
-                ("valoper1".to_string(), Uint128::from(20u128)),
-                ("valoper3".to_string(), Uint128::from(80u128))
-            ]
-        );
-    }
+    assert!(withdraw.is_zero());
```
Comment (albertandrejev) [inline contracts/distribution/src/contract.rs:175]: is it ok to throw panic here?

File: integration_tests/src/testcases/distribution.test.ts
Comment (albertandrejev) [inline integration_tests/src/testcases/distribution.test.ts:1]: Can you add integration test with on_top?

Comment (foxpy) [reply integration_tests/src/testcases/distribution.test.ts:1]: I added on_top into a few existing tests, but I didn't copy all existing unit tests into integration ones, that would have been overkill

**Review by albertandrejev — APPROVED** [review]

**Review by ratik — APPROVED** [review]

**Pull request merged**